intake_nested_yaml_catalog/nested_yaml_catalog.py

```diff
@@ -0,0 +1,186 @@
+        """Return YAML representation of this data-source
+
+        The output may be roughly appropriate for inclusion in a YAML
+        catalog. This is a best-effort implementation
+
+        Parameters
+        ----------
+        with_plugin: bool
+            If True, create a "plugins" section, for cases where this source
+            is created with a plugin not expected to be in the global Intake
+            registry.
+        """
+        from yaml import dump
+        data = self._yaml(with_plugin=with_plugin)
+
+        # convert back to syntactic sugar of yaml_nested_cat
+        yaml_nested_cat = to_yaml_nested_cat(data, self.__module__ + "." + self.__class__.__name__)
+        return dump(yaml_nested_cat, default_flow_style=False)
+
+
+global_registry['yaml_cat'] = YAMLCatalog
+
+
+class YAMLFileNestedCatalog(YAMLFileCatalog):
+    """
+    Catalog as described by a single YAML file with hierarchy. This is syntactic sugar of `YAMLCatalog`.
+    Example:
+    ```
+      sources:
+        entity:
+          property:
+            clean_attributes:
+              args:
+                urlpath: s3://foo
+              driver: parquet
+            zestimate:
+              args:
+                urlpath: '{{ CATALOG_DIR }}/zestimate.csv'
+              description: foo
+              driver: csv
+    ```
+
+    Clean_attributes can be accessed as such:
+    >>> catalog.entity.property.clean_attributes.describe()
+    """
+    version = __version__
+    container = 'catalog'
+    partition_access = None
+    name = 'yaml_nested_cat'
+
+    def __init__(self, path, autoreload=True, **kwargs):
+        """
+        Parameters
+        ----------
+        path: str
+            Location of the file to parse (can be remote)
+        reload : bool
+            Whether to watch the source file for changes; make False if you want
+            an editable Catalog
+        """
+        self.path = path
+        self.text = None
+        self.autoreload = autoreload  # set this to False if don't want reloads
+        super(YAMLFileNestedCatalog, self).__init__(path, autoreload, **kwargs)
+
+    def parse(self, text):
+        self.text = text
+        data = yaml_load(self.text)
+
+        try:
+            # Try the default YAMLFileCatalog first
+            # Reuse default YAMLFileCatalog YAML parser
+            # parse() does the heavy lifting of populating the catalog
+            super().parse(text)
+        except ValidationError:
+            # Try to parse as a nested Catalog by
+            # transforming it to above yaml_cat expected format.
+            transformed_data = to_yaml_cat(data)
+            transformed_text = yaml.dump(transformed_data, default_flow_style=False)
+            super().parse(transformed_text)
+
+
+# Override yaml_file_cat to support a catalog with taxonomy
+# yet fallback to YAMLFileCatalog if the file is in that format already
+global_registry['yaml_file_cat'] = YAMLFileNestedCatalog
```

Comment:
How would the new plugin entry be the default for `intake.Catalog("cat.yaml")`?

Reply:
It would not be, but would show up as `intake.open_hierarchical_yaml`, or `intake_zillow.open_catalog` (depending on the driver's name and how you choose to structure your repo). 

Maybe eventually it can take over, or there can be something in the top-level "metadata" key that marks the syntax, but I wouldn't like to swap out such a basic component yet, even if it only happens when this package is installed. I can imagine that swap leading to a lot of confusion.

Reply:
Instead of swapping the `yaml_file_cat` component, how about instead a PR to update `yaml_file_cat` to support nested hierarchies? 

That'd achieve two purposes:
1. Users of this plugin (and internal Zillow users) wouldn't be confused between `intake.open_yaml_cat()` and `intake.Catalog(path)` and `intake.open_yaml_nested_cat()`
2. Encourage the community to organize their datasets and avoid a data swamp.

Reply:
I see, I could provide user friendly functions via an intake-zillow (or intake-dal) plugin that exposes `intake_dal.Catalog(path), intake_dal.open_catalog()`.  What are your thoughts on providing this hierarchical YAML support within intake-core vs a plugin?

This intake-zillow/intake-dal plugin would provide helper functions for nested catalogs (and the dataset consistent access across storage system environments).

Reply:
Fixed!  Thank you for catching this hack!

I'm writing another plugin that can be invoked using `intake.DalCatalog("cat.yaml")` which would inherit from `YAMLFileNestedCatalog` and add the ability to abstracts a dataset over heterogenous storage systems.

Example entry:
```yaml
    my_datasource:
      driver: dal
      metadata:
        storage:
          default: 'local'
          local: 'csv://{{ CATALOG_DIR }}/data/user_events.csv'
          batch: 'parquet://{{ CATALOG_DIR }}/data/user_events.parquet'
```

`DalCatalog` plugin would do similar YAML transform but to:
```yaml
      my_datasource:
       driver:
         local:
           default: true
           class: intake.source.csv.CSVSource
           args:
             urlpath: '{{ CATALOG_DIR }}/data/user_events.csv'
         batch:
           class: intake_parquet.source.ParquetSource
           args:
             urlpath: '{{ CATALOG_DIR }}/data/user_events.parquet'
```

Reply:
I think we should talk about this, there are many extension points here. The latter version, with multiple drivers, is already possible, and choices between arguments could already be done via user parameters or the alias class. I recommend you have a look at [that](https://github.com/intake/intake/blob/master/intake/source/base.py#L379) for some ideas of how I was thinking about this kind of problem before.

Reply:
`AliasSource` is very cool!  Wouldn't usig `AliasSource` however require new entries in the Catalog possibly confusing users?

```yaml
    my_datasource:
      driver: intake.source.base.AliasSource
      args:
       target: "{{mode}}"
       mapping:
         local: my_datasource_local
         batch: my_datasource_batch
      parameters:
        mode:
          description: which to alias
          type: str
          default: local
          allowed: ["local", "batch"]
    my_datasource_local:
      driver: csv
      args:
        urlpath: '{{ CATALOG_DIR }}/data/user_events.csv'
    my_datasource_batch:
      driver: parquet
      args:
        urlpath: '{{ CATALOG_DIR }}/data/user_events.parquet'
```

Reply:
I think I see, you're suggesting to use the same pattern as `AliasSource` to front and specialize the source at access time!

Reply:
Exactly so. There are, in fact, *many* patterns you could use, which is a benefit and perhaps drawback of an open system like Intake.